### SETUP

In [1]:
import pandas as pd
import numpy as np

/Users/hacktiv8/anaconda3/envs/data_science/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


# Missing Values

## Pandas Missing Values

In pandas, missing values are represented by `NaN` for numeric data and `None` for non-numeric data (i.e., objects, strings). These are considered as the standard missing value representations. Let's explore how pandas handles standard and non-standard missing values:

### Standard Missing Values

#### **NaN**

In pandas, the `NaN` (Not a Number) value is the default representation for missing numeric data. It is a special floating-point value used to indicate missing or undefined data. Pandas automatically handles operations involving `NaN`, propagating it through calculations without raising errors. For example:

In [2]:
data = {'A': [1, 2, np.nan, 4, 5],
        'B': [10, 20, 30, np.nan, 50]}
df = pd.DataFrame(data)

df

A     B
0  1.0  10.0
1  2.0  20.0
2  NaN  30.0
3  4.0   NaN
4  5.0  50.0

In [3]:
print(df.mean())

A     3.0
B    27.5
dtype: float64


In the above example, the `NaN` values are propagated through the mean calculation, and the result is 3.0 for column 'A' and 27.5 for column 'B'.

#### **None**

In pandas, for non-numeric data, the `None` value is used to represent missing values. It is a Python object representing the absence of a value. Pandas automatically converts Python `None` to `NaN` when working with non-numeric data. For example:

In [4]:
data = {'Name': ['Alice', 'Bob', None, 'Eve'],
        'Age': [25, None, 30, 35]}
df = pd.DataFrame(data)

df

Name   Age
0  Alice  25.0
1    Bob   NaN
2   None  30.0
3    Eve  35.0

In this example, 'None' is automatically converted to `NaN` in the 'Age' column.

### Non-Standard Missing Values

Non-standard missing values refer to specific representations used by data sources to indicate missing data. These values are not automatically converted to `NaN` or `None` by pandas, so you need to handle them explicitly during data exploration and preprocessing.

For example, some datasets might use "N/A," "NA," "UNKNOWN," or other special values to indicate missing data. When loading such data into pandas, you need to specify those non-standard missing values as `na_values` during data import:

In [5]:
data = pd.DataFrame({'Column1': [1, 2, 'N/A', 4, 5],
        'Column2': [10, 'NA', 30, 40, 50]})
data

Column1 Column2
0       1      10
1       2      NA
2     N/A      30
3       4      40
4       5      50

In [6]:
data.to_csv('data.csv')

In [7]:
df = pd.read_csv('data.csv', na_values=['N/A', 'NA'])
df

Unnamed: 0  Column1  Column2
0           0      1.0     10.0
1           1      2.0      NaN
2           2      NaN     30.0
3           3      4.0     40.0
4           4      5.0     50.0

In this example, "N/A" and "NA" are treated as missing values and converted to `NaN` during data import.

## How to Check Missing Values in Pandas

You can use the `isnull()` method to check for missing values in pandas. The `isnull()` method returns a DataFrame of the same shape as the original, where each cell contains a boolean value indicating whether the corresponding value is missing (**True**) or not missing (**False**).

In [8]:
df = pd.DataFrame({
                  'Order_ID': [101, 102, 103, 104, 105],
                  'Customer_Name': ['Alice', 'Bob', 'Charlie', None, 'Eve'],
                  'Product': ['Laptop', 'Keyboard', 'Monitor', 'Mouse', None],
                  'Quantity': [1, 2, np.nan, 4, 5],
                  'Price': [1000, 500, np.nan, 200, 300]
              })
df

Order_ID Customer_Name   Product  Quantity   Price
0       101         Alice    Laptop       1.0  1000.0
1       102           Bob  Keyboard       2.0   500.0
2       103       Charlie   Monitor       NaN     NaN
3       104          None     Mouse       4.0   200.0
4       105           Eve      None       5.0   300.0

In [9]:
df.isnull()

Order_ID  Customer_Name  Product  Quantity  Price
0     False          False    False     False  False
1     False          False    False     False  False
2     False          False    False      True   True
3     False           True    False     False  False
4     False          False     True     False  False

### Exploring Missing Values

You can use the `info()` method to get a concise summary of the DataFrame, including the count of non-null values in each column. By comparing the count of non-null values with the total number of rows, you can easily identify columns with missing values.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       5 non-null      int64  
 1   Customer_Name  4 non-null      object 
 2   Product        4 non-null      object 
 3   Quantity       4 non-null      float64
 4   Price          4 non-null      float64
dtypes: float64(2), int64(1), object(2)
memory usage: 328.0+ bytes


In the output, the "Non-Null Count" column indicates the number of non-null values in each column. Columns with fewer non-null values than the total number of rows have missing values.

### Counting Missing Values

To count the number of missing values in each column, you can use the isnull() method followed by sum().

In [11]:
df.isnull().sum()

Order_ID         0
Customer_Name    1
Product          1
Quantity         1
Price            1
dtype: int64

In [12]:
print('Total missing values in data:',df.isnull().sum().sum())

Total missing values in data: 4


## How to Handling Missing Values in Pandas

Pandas provides several methods to handle missing values. Some common techniques include:

### Dropping Rows or Columns with Missing Values

You can use the `dropna()` method to remove rows or columns with missing values.

In [13]:
# Remove rows with any missing values
df_cleaned = df.dropna()

# Remove columns with any missing values
df_cleaned = df.dropna(axis=1)

### Filling Missing Values

You can use the `fillna()` method to fill missing values with specific values.

In [14]:
# Fill missing values in the 'Quantity' and 'Price' columns with the mean
df_filled = df.fillna({'Quantity': df['Quantity'].mean(), 'Price': df['Price'].mean()})

### Interpolation

You can use interpolation methods, such as linear or polynomial, to fill missing values based on existing data.

In [15]:
# Use linear interpolation to fill missing values
df_interpolated = df.interpolate()
df_interpolated

/var/folders/80/9nz19ndj3130jx_8t5hgzwv80000gn/T/ipykernel_11954/980070153.py:2: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df_interpolated = df.interpolate()


Order_ID Customer_Name   Product  Quantity   Price
0       101         Alice    Laptop       1.0  1000.0
1       102           Bob  Keyboard       2.0   500.0
2       103       Charlie   Monitor       3.0   350.0
3       104          None     Mouse       4.0   200.0
4       105           Eve      None       5.0   300.0

In [16]:
df #NaN in Quantity and Price imputed by 3.0 and 350.0

Order_ID Customer_Name   Product  Quantity   Price
0       101         Alice    Laptop       1.0  1000.0
1       102           Bob  Keyboard       2.0   500.0
2       103       Charlie   Monitor       NaN     NaN
3       104          None     Mouse       4.0   200.0
4       105           Eve      None       5.0   300.0

### Replace Missing Values with Specific Values

You can use the replace() method to replace specific missing values with custom values.

In [17]:
# Replace 'None' with 'Unknown' in column Product, row-4
df_replaced = df.replace(to_replace=[None], value='Unknown')
df_replaced


Order_ID Customer_Name   Product  Quantity   Price
0       101         Alice    Laptop       1.0  1000.0
1       102           Bob  Keyboard       2.0   500.0
2       103       Charlie   Monitor       NaN     NaN
3       104       Unknown     Mouse       4.0   200.0
4       105           Eve   Unknown       5.0   300.0

# Data Duplicate

Let's generate a sample DataFrame with duplicate rows in an e-commerce dataset:

In [18]:
# Sample data with duplicate rows
data = {
    'Order_ID': [101, 102, 103, 103, 104],
    'Customer_Name': ['Alice', 'Bob', 'Charlie', 'Charlie', 'Eve'],
    'Product': ['Laptop', 'Keyboard', 'Monitor', 'Monitor', 'Mouse'],
    'Quantity': [1, 2, 1, 1, 1],
    'Price': [1000, 500, 300, 300, 200]
}

df = pd.DataFrame(data)
df


Order_ID Customer_Name   Product  Quantity  Price
0       101         Alice    Laptop         1   1000
1       102           Bob  Keyboard         2    500
2       103       Charlie   Monitor         1    300
3       103       Charlie   Monitor         1    300
4       104           Eve     Mouse         1    200

In this example DataFrame, we have duplicate rows for Order_ID 103, as the customer "Charlie" ordered the same product "Monitor" twice.

## Check Data Duplicate in Pandas
To check for duplicate rows in a DataFrame, you can use the `duplicated()` method, which returns a boolean Series indicating whether each row is a duplicate (True) or not (False).

In [19]:
duplicates = df.duplicated()
duplicates

0    False
1    False
2    False
3     True
4    False
dtype: bool

## Remove Data Duplicate
To remove duplicate rows from a DataFrame, you can use the `drop_duplicates()` method. By default, it keeps the first occurrence of each duplicate row and removes the subsequent duplicates.

In [20]:
df_cleaned = df.drop_duplicates()
df_cleaned

Order_ID Customer_Name   Product  Quantity  Price
0       101         Alice    Laptop         1   1000
1       102           Bob  Keyboard         2    500
2       103       Charlie   Monitor         1    300
4       104           Eve     Mouse         1    200

The duplicate row with 'Order_ID' 102 and 'Customer_Name' 'Bob' is removed, and only the first occurrence is kept.

You can customize the behavior using the keep parameter:
- **keep='first'**: (default) Keeps the first occurrence of each duplicate row and removes the subsequent duplicates.
- **keep='last'**: Keeps the last occurrence of each duplicate row and removes the preceding duplicates.
- **keep=False**: Removes all occurrences of duplicate rows.



In [21]:
df_cleaned_last = df.drop_duplicates(keep='last')
df_cleaned_last

Order_ID Customer_Name   Product  Quantity  Price
0       101         Alice    Laptop         1   1000
1       102           Bob  Keyboard         2    500
3       103       Charlie   Monitor         1    300
4       104           Eve     Mouse         1    200

The duplicate row with 'Order_ID' 102 and 'Customer_Name' 'Bob' is removed, but the last occurrence is kept.

## Subset for Duplicates
You can check for duplicates based on a subset of columns using the subset parameter. This allows you to identify duplicates based on specific columns rather than the entire row.

In [22]:
duplicates_subset = df.duplicated(subset=['Order_ID', 'Customer_Name'])

duplicates_subset

0    False
1    False
2    False
3     True
4    False
dtype: bool

In this example, the last row (index 4) is a duplicate based on the 'Order_ID' and 'Customer_Name' columns.

## Handling Duplicates Based on Key Columns:
To remove duplicates based on a subset of columns, you can use the `drop_duplicates()` method with the `subset` parameter.

In [23]:
# Remove duplicates based on the 'Order_ID' and 'Customer_Name' columns, keeping the last occurrence
df_cleaned_subset = df.drop_duplicates(subset=['Order_ID', 'Customer_Name'], keep='last')

df_cleaned_subset

Order_ID Customer_Name   Product  Quantity  Price
0       101         Alice    Laptop         1   1000
1       102           Bob  Keyboard         2    500
3       103       Charlie   Monitor         1    300
4       104           Eve     Mouse         1    200

In this example, the duplicate row with 'Order_ID' 102 and 'Customer_Name' 'Bob' is removed, and only the last occurrence is kept.

# Incosistent Data Handling

## Data Type Conversion

Data type conversion in pandas using the `astype()` method allows you to explicitly change the data type of one or more columns in a DataFrame. This operation is crucial because sometimes the imported data may not have the correct data types, or you might want to convert data for better analysis and processing. You can use `astype()` to convert columns to numeric, datetime, string, or any other appropriate data type.

In [24]:
# Sample sales transaction data
data = {
    'Order_ID': [101, 102, 103, 104, 105],
    'Date': ['2023-07-15', '2023-07-16', '2023-07-17', '2023-07-18', '2023-07-19'],
    'Customer_ID': ['C001', 'C002', 'C003', 'C004', 'C005'],
    'Product': ['Laptop', 'Keyboard', 'Monitor', 'Mouse', 'Headphones'],
    'Quantity': ['1', '2', '3', '4', '5'],
    'Price': ['1000.50', '250.25', '350.75', '75.20', '120.99']
}

df = pd.DataFrame(data)
df

Order_ID        Date Customer_ID     Product Quantity    Price
0       101  2023-07-15        C001      Laptop        1  1000.50
1       102  2023-07-16        C002    Keyboard        2   250.25
2       103  2023-07-17        C003     Monitor        3   350.75
3       104  2023-07-18        C004       Mouse        4    75.20
4       105  2023-07-19        C005  Headphones        5   120.99

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Order_ID     5 non-null      int64 
 1   Date         5 non-null      object
 2   Customer_ID  5 non-null      object
 3   Product      5 non-null      object
 4   Quantity     5 non-null      object
 5   Price        5 non-null      object
dtypes: int64(1), object(5)
memory usage: 368.0+ bytes


In this example DataFrame, the 'Date', 'Quantity', and 'Price' columns are stored as strings. We need to convert them to appropriate data types.

The 'Date' column should be converted to the datetime data type for better time-related operations.

In [26]:
df['Date'] = pd.to_datetime(df['Date'])
print(df.dtypes)

Order_ID                int64
Date           datetime64[ns]
Customer_ID            object
Product                object
Quantity               object
Price                  object
dtype: object


'Quantity' represents numeric values, so we should convert it to the integer data type.

In [27]:
df['Quantity'] = df['Quantity'].astype(int)
print(df.dtypes)

Order_ID                int64
Date           datetime64[ns]
Customer_ID            object
Product                object
Quantity                int64
Price                  object
dtype: object


'Price' represents currency values, so we should convert it to the float data type.

In [28]:
df['Price'] = df['Price'].astype(float)
print(df.dtypes)

Order_ID                int64
Date           datetime64[ns]
Customer_ID            object
Product                object
Quantity                int64
Price                 float64
dtype: object


## String

In [29]:
df = pd.DataFrame({
    'Transaction_ID': [101, 102, 103, 104, 105],
    'Customer_Name': ['  John Doe  ', 'Alice', '  Bob  ', 'Charlie', 'Eve'],
    'Product_Description': ['Laptop Asus Core i5', 'Keyboard Logitech', 'Monitor Samsung', 'Mouse HP', 'Headphones Sony']
})

df

Transaction_ID Customer_Name  Product_Description
0             101    John Doe    Laptop Asus Core i5
1             102         Alice    Keyboard Logitech
2             103         Bob        Monitor Samsung
3             104       Charlie             Mouse HP
4             105           Eve      Headphones Sony

**1. Removing Leading/Trailing Whitespaces**:

You can remove leading and trailing whitespaces from string columns using the `str.strip()` method:

In [30]:
# Remove leading/trailing whitespaces from 'Customer_Name' and 'Product_Description'
df['Customer_Name'] = df['Customer_Name'].str.strip()
df['Product_Description'] = df['Product_Description'].str.strip()
df

Transaction_ID Customer_Name  Product_Description
0             101      John Doe  Laptop Asus Core i5
1             102         Alice    Keyboard Logitech
2             103           Bob      Monitor Samsung
3             104       Charlie             Mouse HP
4             105           Eve      Headphones Sony

**2. Changing Case**:

You can change the case of string columns using the `str.lower()` or `str.upper()` methods:

In [31]:
# Convert 'Customer_Name' to lowercase
df['Customer_Name'] = df['Customer_Name'].str.lower()

# Convert 'Product_Description' to uppercase
df['Product_Description'] = df['Product_Description'].str.upper()

df

Transaction_ID Customer_Name  Product_Description
0             101      john doe  LAPTOP ASUS CORE I5
1             102         alice    KEYBOARD LOGITECH
2             103           bob      MONITOR SAMSUNG
3             104       charlie             MOUSE HP
4             105           eve      HEADPHONES SONY

**3. Removing or Replacing Substrings**:

You can remove or replace specific substrings in string columns using the `str.replace()` method:

In [32]:
# Remove 'Core i5' from 'Product_Description'
df['Product_Description'] = df['Product_Description'].str.replace('Core i5', '')

# Replace 'Sony' with 'Sony Corporation' in 'Product_Description'
df['Product_Description'] = df['Product_Description'].str.replace('Sony', 'Sony Corporation')

df

Transaction_ID Customer_Name  Product_Description
0             101      john doe  LAPTOP ASUS CORE I5
1             102         alice    KEYBOARD LOGITECH
2             103           bob      MONITOR SAMSUNG
3             104       charlie             MOUSE HP
4             105           eve      HEADPHONES SONY

**4. Splitting and Extracting Substrings**:

You can split and extract substrings from string columns using the `str.split()` and indexing:

In [33]:
# Split 'Product_Description' into two separate columns
# df['Product_Description'].str.split(' ', 1, expand=True) deprecated
df['Product_Description'].str.split(' ', n=1, expand=True)

0             1
0      LAPTOP  ASUS CORE I5
1    KEYBOARD      LOGITECH
2     MONITOR       SAMSUNG
3       MOUSE            HP
4  HEADPHONES          SONY

In [34]:
df[['Product_Brand', 'Product_Model']] = df['Product_Description'].str.split(' ', n=1, expand=True)
df

Transaction_ID Customer_Name  Product_Description Product_Brand  \
0             101      john doe  LAPTOP ASUS CORE I5        LAPTOP   
1             102         alice    KEYBOARD LOGITECH      KEYBOARD   
2             103           bob      MONITOR SAMSUNG       MONITOR   
3             104       charlie             MOUSE HP         MOUSE   
4             105           eve      HEADPHONES SONY    HEADPHONES   

  Product_Model  
0  ASUS CORE I5  
1      LOGITECH  
2       SAMSUNG  
3            HP  
4          SONY

## Datetime

In [35]:
data = {
    'Order_ID': [101, 102, 103, 104, 105],
    'Date_Time': ['2023-07-15 10:30:00', '2023-07-16 15:45:00', '2023-07-17 12:00:00', '2023-07-18 09:00:00', '2023-07-19 16:20:00'],
    'Customer_ID': ['C001', 'C002', 'C003', 'C004', 'C005'],
    'Product': ['Laptop', 'Keyboard', 'Monitor', 'Mouse', 'Headphones'],
    'Quantity': [1, 2, 3, 4, 5],
    'Price': [1000, 250, 350, 75, 120]
}

df = pd.DataFrame(data)
df['Date_Time'] = pd.to_datetime(df['Date_Time'])

df

Order_ID           Date_Time Customer_ID     Product  Quantity  Price
0       101 2023-07-15 10:30:00        C001      Laptop         1   1000
1       102 2023-07-16 15:45:00        C002    Keyboard         2    250
2       103 2023-07-17 12:00:00        C003     Monitor         3    350
3       104 2023-07-18 09:00:00        C004       Mouse         4     75
4       105 2023-07-19 16:20:00        C005  Headphones         5    120

**1. Extracting Date Components**:

You can extract various date components from the 'Date_Time' column, such as year, month, day, hour, minute, and second, using the `dt` accessor.

In [36]:
# Extract year, month, and day from 'Date_Time'
df['Year'] = df['Date_Time'].dt.year
df['Month'] = df['Date_Time'].dt.month
df['Day'] = df['Date_Time'].dt.day

df[['Date_Time', 'Year', 'Month', 'Day']]

Date_Time  Year  Month  Day
0 2023-07-15 10:30:00  2023      7   15
1 2023-07-16 15:45:00  2023      7   16
2 2023-07-17 12:00:00  2023      7   17
3 2023-07-18 09:00:00  2023      7   18
4 2023-07-19 16:20:00  2023      7   19

**2. Formatting Datetime**:

You can format the 'Date_Time' column into a specific datetime format using the `dt.strftime()` method.

In [37]:
# Format 'Date_Time' to a specific format (e.g., "yyyy-mm-dd HH:MM")
df['Formatted_Date'] = df['Date_Time'].dt.strftime('%Y-%m-%d %H:%M')

df[['Date_Time', 'Formatted_Date']]

Date_Time    Formatted_Date
0 2023-07-15 10:30:00  2023-07-15 10:30
1 2023-07-16 15:45:00  2023-07-16 15:45
2 2023-07-17 12:00:00  2023-07-17 12:00
3 2023-07-18 09:00:00  2023-07-18 09:00
4 2023-07-19 16:20:00  2023-07-19 16:20

**3. Calculating Time Differences**:

You can calculate the time differences between two datetime columns, such as the difference between the 'Date_Time' and the current time.

In [38]:
# Calculate time difference from the current time
current_time = pd.to_datetime('2023-07-20 10:00:00')
df['Time_Difference'] = current_time - df['Date_Time']

df[['Date_Time', 'Time_Difference']]

Date_Time Time_Difference
0 2023-07-15 10:30:00 4 days 23:30:00
1 2023-07-16 15:45:00 3 days 18:15:00
2 2023-07-17 12:00:00 2 days 22:00:00
3 2023-07-18 09:00:00 2 days 01:00:00
4 2023-07-19 16:20:00 0 days 17:40:00

**4. Extracting Time Intervals**:

You can extract time intervals, such as hours and minutes, from the 'Date_Time' column.

In [39]:
# Extract hours and minutes from 'Date_Time'
df['Hour'] = df['Date_Time'].dt.hour
df['Minute'] = df['Date_Time'].dt.minute

df[['Date_Time', 'Hour', 'Minute']]

Date_Time  Hour  Minute
0 2023-07-15 10:30:00    10      30
1 2023-07-16 15:45:00    15      45
2 2023-07-17 12:00:00    12       0
3 2023-07-18 09:00:00     9       0
4 2023-07-19 16:20:00    16      20




**5. Handling Timezones**:

Pandas provides support for handling timezones in datetime data. You can convert the timezone of the 'Date_Time' column using the `dt.tz_localize()` method.


In [40]:
# Assuming the original datetime is in UTC and we want to convert it to Jakarta timezone (Asia/Jakarta)
df['Date_Time_Jakarta'] = df['Date_Time'].dt.tz_localize('UTC').dt.tz_convert('Asia/Jakarta')

df[['Date_Time', 'Date_Time_Jakarta']]

Date_Time         Date_Time_Jakarta
0 2023-07-15 10:30:00 2023-07-15 17:30:00+07:00
1 2023-07-16 15:45:00 2023-07-16 22:45:00+07:00
2 2023-07-17 12:00:00 2023-07-17 19:00:00+07:00
3 2023-07-18 09:00:00 2023-07-18 16:00:00+07:00
4 2023-07-19 16:20:00 2023-07-19 23:20:00+07:00

In this example, we converted the 'Date_Time' column from UTC to the Jakarta timezone (Asia/Jakarta).

# Python Scripting for Data Processing Steps

You've already learn how to perform data wrangling in Pandas. Now, we will do data manipulation practice but in different way.

Imagine that you can't work using Google Colab or Jupyter Notebook. Instead, you use Python script file in order to make a automated data manipulation/preprocessing pipeline.

Our goal in this section is to make an excel file that consist of several sheet from query and grouping

Our expectation:

```
data.xlsx
 ├── Complete
 ├── Online Channel
 ├── Offline Channel
 └── Countries Summary
```

First of all, let see the data:

In [61]:
dat = pd.read_csv('https://github.com/FTDS-learning-materials/phase-0/raw/main/src/travel%20insurance.csv')

dat.head()

Agency    Agency Type Distribution Channel                     Product Name  \
0    CBH  Travel Agency              Offline               Comprehensive Plan   
1    CBH  Travel Agency              Offline               Comprehensive Plan   
2    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   
3    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   
4    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   

  Claim  Duration Destination  Net Sales  Commision (in value) Gender  Age  
0    No       186    MALAYSIA      -29.0                  9.57      F   81  
1    No       186    MALAYSIA      -29.0                  9.57      F   71  
2    No        65   AUSTRALIA      -49.5                 29.70    NaN   32  
3    No        60   AUSTRALIA      -39.6                 23.76    NaN   32  
4    No        79       ITALY      -19.8                 11.88    NaN   41

To make an automated data cleaning and preprocessing, we need to explore our data in aims to decide which steps that we will do next.

In [62]:
dat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63326 entries, 0 to 63325
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Agency                63326 non-null  object 
 1   Agency Type           63326 non-null  object 
 2   Distribution Channel  63326 non-null  object 
 3   Product Name          63326 non-null  object 
 4   Claim                 63326 non-null  object 
 5   Duration              63326 non-null  int64  
 6   Destination           63326 non-null  object 
 7   Net Sales             63326 non-null  float64
 8   Commision (in value)  63326 non-null  float64
 9   Gender                18219 non-null  object 
 10  Age                   63326 non-null  int64  
dtypes: float64(2), int64(2), object(7)
memory usage: 5.3+ MB


Based on the summary info, there are many rows that missing in Gender column. We will take out the column from the data and then we will rename `Commision (in value)` into `Commision`. We will perform them directly in the script.

However, we need to preprocess the data in aims to store in each sheet that we need using query or grouping. Then, we will try to explore the codes based on our needs.

In [43]:
from IPython.core.display import display, HTML

display(HTML('<h3><b>Online Channel</b></h3><br>'))
ol_ch = dat[dat['Distribution Channel']=='Online']
display(ol_ch.head())

display(HTML('<h3><b>Offline Channel</b></h3><br>'))
of_ch = dat[dat['Distribution Channel']=='Offline']
display(of_ch.head())

/var/folders/80/9nz19ndj3130jx_8t5hgzwv80000gn/T/ipykernel_11954/2706172112.py:1: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


Agency    Agency Type Distribution Channel                     Product Name  \
2    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   
3    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   
4    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   
5    JZI       Airlines               Online                       Value Plan   
6    CWT  Travel Agency               Online  Rental Vehicle Excess Insurance   

  Claim  Duration    Destination  Net Sales  Commision (in value) Gender  Age  
2    No        65      AUSTRALIA      -49.5                 29.70    NaN   32  
3    No        60      AUSTRALIA      -39.6                 23.76    NaN   32  
4    No        79          ITALY      -19.8                 11.88    NaN   41  
5    No        66  UNITED STATES     -121.0                 42.35      F   44  
6    No        47  UNITED STATES      -39.6                 23.76    NaN   32

Agency    Agency Type Distribution Channel        Product Name Claim  \
0      CBH  Travel Agency              Offline  Comprehensive Plan    No   
1      CBH  Travel Agency              Offline  Comprehensive Plan    No   
400    CSR  Travel Agency              Offline  Comprehensive Plan    No   
401    CSR  Travel Agency              Offline  Comprehensive Plan    No   
402    CCR  Travel Agency              Offline  Comprehensive Plan    No   

     Duration Destination  Net Sales  Commision (in value) Gender  Age  
0         186    MALAYSIA      -29.0                  9.57      F   81  
1         186    MALAYSIA      -29.0                  9.57      F   71  
400        92    MALAYSIA       29.0                  9.57      F   72  
401        92    MALAYSIA       29.0                  9.57      F   77  
402         5    MALAYSIA       29.0                  9.57      M  118

In [ ]:
display(HTML('<h3><b>Countries Summary</b></h3><br>'))

summary = dat.groupby(['Destination','Agency Type','Distribution Channel','Claim']).agg({'Product Name':'count',
                                                                                         'Duration':'mean',
                                                                                         'Net Sales':'max',
                                                                                         'Commision (in value)':'sum',
                                                                                         'Age':'mean'}).reset_index()
summary.head()

Destination    Agency Type Distribution Channel Claim  Product Name  \
0     ALBANIA  Travel Agency               Online    No             1   
1      ANGOLA  Travel Agency               Online    No             1   
2   ARGENTINA  Travel Agency               Online    No            21   
3   ARGENTINA  Travel Agency               Online   Yes             1   
4     ARMENIA  Travel Agency               Online    No             1   

     Duration  Net Sales  Commision (in value)        Age  
0   28.000000       80.0                  0.00  36.000000  
1    2.000000        0.0                 16.00  19.000000  
2   35.904762      114.0                237.35  36.238095  
3   30.000000       79.2                 47.52  30.000000  
4  232.000000       25.0                  0.00  36.000000

Finally, here is your code:

```py
import pandas as pd
import sys

class DataProcessor:
    def __init__(self, file_path):
        self.file_path = file_path
        self.df = None

    def load_data(self):
        self.df = pd.read_csv(self.file_path).reset_index()

    def clean_data(self):

        #Rename Commision (in value) Column
        self.df.rename({'Commision (in value)':'Commision'},axis=1,inplace=True)

        #Take out Gender Column
        self.df.drop(columns='Gender',inplace=True)

        # Drop duplicates
        self.df.drop_duplicates(inplace=True)
    
    def simple_query(self, column, value):
        return self.df[self.df[column]==value].drop(columns='index')

    def summary(self,cols=[],aggs={}):
        self.tmp = self.df.groupby(cols).agg(aggs).reset_index()
        self.renames = dict()
        for col,agg in aggs.items():
            self.renames[col] = f'{agg} {col}'.title()

        return self.tmp.rename(self.renames,axis=1)


    def save_data(self, output_file_path, data, sheet_names):
        with pd.ExcelWriter(output_file_path) as writer:
            self.df.to_excel(writer, sheet_name='Complete', index=False)
            for i,dat in enumerate(data):
              try:
                dat.to_excel(writer, sheet_name=sheet_names[i], index=False)
              except:
                display(dat)
                print(i)


if __name__ == "__main__":
    input_file = sys.argv[2]
    output_file = sys.argv[4]

    print("Data Processing Script Started...")

    # Create an instance of DataCleaning
    proc = DataProcessor(file_path=input_file)

    # Load data
    print("Loading data...")

    try:
      proc.load_data()
    except Exception as e:
      print('Error has occured while loading the data')
      print(e)
      print('Cannot be Continued')
      print('Stop')
      sys.exit()
      

    # Clean data
    print("Cleaning data...")
    proc.clean_data()

    # Preprocess data
    print("Preprocessing data...")
    ol = proc.simple_query('Distribution Channel','Online')
    of = proc.simple_query('Distribution Channel','Offline')
    sum = proc.summary(cols=['Destination','Agency Type','Distribution Channel','Claim'],
              aggs={'index':'count','Duration':'mean','Net Sales':'max','Commision':'sum','Age':'mean'})

    # Save cleaned data
    print("Saving cleaned data...")

    proc.save_data(output_file,data=[ol,of,sum], sheet_names=['Online','Offline','Countries Summary'])

    print("Data Processing Script Completed!")
```

You will see on the code, below ```if __name__ = '__main__'``` there are lines:

```
input_file = sys.argv[2]
output_file = sys.argv[4]
```

and also `sys` library is imported.

So, for what? as you learned from the slide, `sys.argv` is used for giving command-line arguments as flags. They resemble to `**kwargs` in Python function.

We will simulate what if we run a simple python script and input the flags and also the values (remember, flag is like argument in function).

In [64]:
with open('test.py', 'w') as f:
  f.write('''
import sys
print(sys.argv)
  ''')

!python test.py -input 'travel insurance.csv' -output clean.xlsx

['test.py', '-input', 'travel insurance.csv', '-output', 'clean.xlsx']


You see that any command after `python` keyword will be read by sys and save to a list in the `sys.argv`.

So, `input_file = sys.argv[2]` means `input_file` values `travel_insurance.csv` and `output_file = sys.argv[4]` means `outut_file` values `clean.xlsx`.

---

To ease your learning, you may don't need to move to vscode or cmd or terminal. Just using this colab to run the code.

We will perform these steps:
* Create data_process.py
* Download data into Google Colab storage
* Running the script



In [63]:
#@title **Creating data_process.py**

with open("data_process.py", "w") as f:
    f.write('''import pandas as pd
import sys

class DataProcessor:
    def __init__(self, file_path):
        self.file_path = file_path
        self.df = None

    def load_data(self):
        self.df = pd.read_csv(self.file_path).reset_index()

    def clean_data(self):

        #Rename Commision (in value) Column
        self.df.rename({'Commision (in value)':'Commision'},axis=1,inplace=True)

        #Take out Gender Column
        self.df.drop(columns='Gender',inplace=True)

        # Drop duplicates
        self.df.drop_duplicates(inplace=True)

    def simple_query(self, column, value):
        return self.df[self.df[column]==value].drop(columns='index')

    def summary(self,cols=[],aggs={}):
        self.tmp = self.df.groupby(cols).agg(aggs).reset_index()
        self.renames = dict()
        for col,agg in aggs.items():
            self.renames[col] = f'{agg} {col}'.title()

        return self.tmp.rename(self.renames,axis=1)


    def save_data(self, output_file_path, data, sheet_names):
        with pd.ExcelWriter(output_file_path) as writer:
            self.df.to_excel(writer, sheet_name='Complete', index=False)
            for i,dat in enumerate(data):
              try:
                dat.to_excel(writer, sheet_name=sheet_names[i], index=False)
              except:
                display(dat)
                print(i)


if __name__ == "__main__":
    input_file = sys.argv[2]
    output_file = sys.argv[4]

    print("Data Processing Script Started...")

    # Create an instance of DataCleaning
    proc = DataProcessor(file_path=input_file)

    # Load data
    print("Loading data...")

    try:
      proc.load_data()
    except Exception as e:
      print('Error has occured while loading the data')
      print(e)
      print('Cannot be Continued')
      print('Stop')
      sys.exit()


    # Clean data
    print("Cleaning data...")
    proc.clean_data()

    # Preprocess data
    print("Preprocessing data...")
    ol = proc.simple_query('Distribution Channel','Online')
    of = proc.simple_query('Distribution Channel','Offline')
    sum = proc.summary(cols=['Destination','Agency Type','Distribution Channel','Claim'],
              aggs={'index':'count','Duration':'mean','Net Sales':'max','Commision':'sum','Age':'mean'})

    # Save cleaned data
    print("Saving cleaned data...")

    proc.save_data(output_file,data=[ol,of,sum], sheet_names=['Online','Offline','Countries Summary'])

    print("Data Processing Script Completed!")
''')

#### **Download travel_insurance.csv**

In [65]:
!wget https://github.com/FTDS-learning-materials/phase-0/raw/main/src/travel%20insurance.csv

--2025-05-08 14:43:23--  https://github.com/FTDS-learning-materials/phase-0/raw/main/src/travel%20insurance.csv
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/FTDS-learning-materials/phase-0/main/src/travel%20insurance.csv [following]
--2025-05-08 14:43:23--  https://raw.githubusercontent.com/FTDS-learning-materials/phase-0/main/src/travel%20insurance.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4570914 (4.4M) [text/plain]
Saving to: ‘travel insurance.csv’

travel insurance.cs 100%[===================>]   4.36M  1.81MB/s    in 2.4s    

2025-05-08 14:43:26 (1.81 MB/s) - ‘travel insur

#### **Running the Script**

In [66]:
!python data_process.py -input 'travel insurance.csv' -output clean.xlsx

/Users/hacktiv8/anaconda3/envs/data_science/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (
Data Processing Script Started...
Loading data...
Cleaning data...
Preprocessing data...
Saving cleaned data...
Data Processing Script Completed!


---

# Additional Materials (Complementary - Self Learning)

## Combining Multiple Datasets

Now, you’ll use `.concat()` to combine city_data with further_city_data. Say you’ve managed to gather some data on two more cities:

In [49]:
city_revenues = pd.Series(
    [4200, 8000, 6500],
    index=["Amsterdam", "Toronto", "Tokyo"]
)
city_employee_count = pd.Series({"Amsterdam": 5, "Tokyo": 8})

further_city_data = pd.DataFrame(
    {"revenue": [7000, 3400], "employee_count": [2, 2]},
    index=["New York", "Barcelona"]
)

city_data = pd.DataFrame({
    "revenue": city_revenues,
    "employee_count": city_employee_count
})

In [50]:
further_city_data

revenue  employee_count
New York      7000               2
Barcelona     3400               2

In [51]:
city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN

This second DataFrame contains info on the cities "New York" and "Barcelona".

You can add these cities to city_data using `.concat()`:

In [52]:
all_city_data = pd.concat([city_data, further_city_data], sort=False)
all_city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN
New York      7000             2.0
Barcelona     3400             2.0

Now, the new variable `all_city_data` contains the values from both DataFrame objects.

By default, `concat()` combines along axis=0. In other words, it appends rows. You can also use it to append columns by supplying the parameter `axis=1`:

In [53]:
city_countries = pd.DataFrame({
    "country": ["Holland", "Japan", "Holland", "Canada", "Spain"],
    "capital": [1, 1, 0, 0, 0]},
    index=["Amsterdam", "Tokyo", "Rotterdam", "Toronto", "Barcelona"]
)

city_countries

country  capital
Amsterdam  Holland        1
Tokyo        Japan        1
Rotterdam  Holland        0
Toronto     Canada        0
Barcelona    Spain        0

In [54]:
cities = pd.concat([all_city_data, city_countries], axis=1, sort=False)
cities

revenue  employee_count  country  capital
Amsterdam   4200.0             5.0  Holland      1.0
Tokyo       6500.0             8.0    Japan      1.0
Toronto     8000.0             NaN   Canada      0.0
New York    7000.0             2.0      NaN      NaN
Barcelona   3400.0             2.0    Spain      0.0
Rotterdam      NaN             NaN  Holland      0.0

Note how Pandas added NaN for the missing values.

If you want to combine only the cities that appear in both DataFrame objects, then you can set the join parameter to inner:

In [55]:
pd.concat([all_city_data, city_countries], axis=1, join="inner")

revenue  employee_count  country  capital
Amsterdam     4200             5.0  Holland        1
Tokyo         6500             8.0    Japan        1
Toronto       8000             NaN   Canada        0
Barcelona     3400             2.0    Spain        0

While it’s most straightforward to combine data based on the index, it’s not the only possibility. You can use `.merge()` to implement a join operation similar to the one from SQL:

In [56]:
countries = pd.DataFrame({
    "population_millions": [17, 127, 37],
    "continent": ["Europe", "Asia", "North America"]
}, index=["Holland", "Japan", "Canada"])

Here, you pass the parameter `left_on="country"` to `.merge()` to indicate what column you want to join on. The result is a bigger DataFrame that contains not only city data, but also the population and continent of the respective countries:

In [57]:
countries

population_millions      continent
Holland                   17         Europe
Japan                    127           Asia
Canada                    37  North America

In [58]:
cities

revenue  employee_count  country  capital
Amsterdam   4200.0             5.0  Holland      1.0
Tokyo       6500.0             8.0    Japan      1.0
Toronto     8000.0             NaN   Canada      0.0
New York    7000.0             2.0      NaN      NaN
Barcelona   3400.0             2.0    Spain      0.0
Rotterdam      NaN             NaN  Holland      0.0

In [59]:
pd.merge(cities, countries, left_on="country", right_index=True)

revenue  employee_count  country  capital  population_millions  \
Amsterdam   4200.0             5.0  Holland      1.0                   17   
Tokyo       6500.0             8.0    Japan      1.0                  127   
Toronto     8000.0             NaN   Canada      0.0                   37   
Rotterdam      NaN             NaN  Holland      0.0                   17   

               continent  
Amsterdam         Europe  
Tokyo               Asia  
Toronto    North America  
Rotterdam         Europe

Note that the result contains only the cities where the country is known and appears in the joined DataFrame.

`.merge()` performs an inner join by default. If you want to include all cities in the result, then you need to provide the how parameter:

In [60]:
pd.merge(
    cities,
    countries,
    left_on="country",
    right_index=True,
    how="left"
)

revenue  employee_count  country  capital  population_millions  \
Amsterdam   4200.0             5.0  Holland      1.0                 17.0   
Tokyo       6500.0             8.0    Japan      1.0                127.0   
Toronto     8000.0             NaN   Canada      0.0                 37.0   
New York    7000.0             2.0      NaN      NaN                  NaN   
Barcelona   3400.0             2.0    Spain      0.0                  NaN   
Rotterdam      NaN             NaN  Holland      0.0                 17.0   

               continent  
Amsterdam         Europe  
Tokyo               Asia  
Toronto    North America  
New York             NaN  
Barcelona            NaN  
Rotterdam         Europe

With this left join, you’ll see all the cities, including those without country data.